core/src/main/scala/kafka/log/Log.scala

```diff
@@ -713,9 +713,19 @@ class Log(@volatile var dir: File,
           }
         } else {
           // we are taking the offsets we are given
-          if (!appendInfo.offsetsMonotonic || appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset)
+          if (!appendInfo.offsetsMonotonic)
             throw new IllegalArgumentException(s"Out of order offsets found in append to $topicPartition: " +
-              records.records.asScala.map(_.offset))
+                                               records.records.asScala.map(_.offset))
+
+          if (appendInfo.firstOrLastOffset < nextOffsetMetadata.messageOffset) {
+            // we may still be able to recover in some situations
+            val firstOrLast = if (appendInfo.firstOffset.isDefined) "First" else "Last"
+            throw new UnexpectedAppendOffsetException(
```

Comment:
Let's document this in the method scaladoc.

Reply:
I think the patch is primarily concerned with the case when the `firstOrLastOffset` is smaller than the current log start offset. I am wondering if we can handle this case here instead of in the caller. All of the logic in `LogManager.truncateFullyAndStartAt` is concerned with making the truncation safe for the cleaner, but if we are writing at the log start offset, then there won't be anything to clean anyway. Additionally, I don't think we actually want to change the log start offset in this case since we want it to match the leader.

I think we also need to be a little careful with the use of `firstOrLastOffset` here. For the new message format, this will be the first offset, so it is fine to use it for the base offset of the new segment that we create for this offset. But for the old message format, it will be the last offset of the batch (if compression is enabled). Maybe we have no choice but to decompress the batch in order to determine the first offset?

Reply:
The logic in `LogManager.truncateFullyAndStartAt` also makes a checkpoint for all logs for the topic partition, which I think we should do, otherwise checkpoint will be inconsistent with actual data. 

We are setting log start offset to the leader's log start offset once we truncate fully and append. It is done in fetcher thread's processPartitionData().

We are not using `firstOrLastOffset` to find the first offset, we are using  `firstOffset` if it is defined. So, the fix is only for the new message format. Do we want to fix that for the old message format as well? The case of setting log start offset in the middle of the batch was introduced with deleteRecords in 1.0 (or 1.1), where we already had new message format -- meaning, that we don't need to backport this to much older versions. I guess it is possible that users could run new versions of kafka with older message format, but is that the case we want to support here?

Reply:
@hachikuji Do you still think we should handle the truncation in append() vs. `LogManager.truncateFullyAndStartAt` given we also need to checkpoint (per my comment above)? Disregard my last paragraph about not addressing message format 1, I will add the code for finding first offset if it is not available.

Reply:
@apovzner Yeah, I was wondering whether the checkpoint was actually required. It could be, but my point was that we do not actually need to change the log start offset when this case is hit. We just need to be able to append a batch which starts a little earlier. In that case, I'm not sure checkpointing is actually needed (the checkpoints are just an optimization anyway).

Ah, I missed the use of `firstOffset`. I think we probably should try to handle the old message format as well since users have tended to upgrade the message format slowly. This should be a rare case, so I don't think decompressing to find the first offset is a big problem.